### PR TITLE
ssl: expose upstream TLS client context config

### PR DESCRIPTION
**Pull request by kyessenov**
Description: Populate certificate validation context for upstream host descriptions. We would like to collect whether upstream has client cert defined for telemetry purposes in a custom filter.

/assign @lizan 

Risk Level: Small
Testing: None, looking for a recommendation how to test this properly.  
Docs Changes: None
Release Notes:

### CODE REVIEW

**Comment by lizan** [comment]
Do you have a WIP PR that is relying on this in the filter? I don’t think this should be in the factory interface but would like to see how you’re using it.

**Comment by kyessenov** [comment]
https://github.com/istio/proxy/pull/2138/files#diff-82ed8ded232725e4efcf543a37d56131R222

Ideally, I could just access the security info from cluster info, but I don't see a way to do it without going through a factory right now.

**Comment by lizan** [comment]
I think ideally you would like to see what is actually being used for the upstream connection? so something like https://github.com/envoyproxy/envoy/pull/6144 but for upstream in StreamInfo sounds more appropriate, no?

**Comment by kyessenov** [comment]
Yes and no. I think we do want an actual connection info for telemetry. But we also want a policy check using the intended connection settings. I realize there is an issue that upstream cluster info is inaccessible during decodeHeaders ATM, but that can be fixed separately.

File: include/envoy/network/transport_socket.h [hunk 1]
```diff
@@ -181,6 +182,11 @@ class TransportSocketFactory {
    */
   virtual bool implementsSecureTransport() const PURE;
 
+  /**
+   * @return CertificateValidationContextConfig the certificate validation context config.
+   */
+  virtual const Ssl::CertificateValidationContextConfig* certificateValidationContext() const PURE;
```
Comment (lizan) [inline include/envoy/network/transport_socket.h]: We need a bit of abstraction for this by not leaking `Ssl` namespace into here, perhaps a Protobuf::Message& config() or abstract class `TransportSocketConfig`. `Ssl::Connection` in `TransportSocket` is due to historical reason and that is to be removed.

Comment (kyessenov) [reply include/envoy/network/transport_socket.h]: It looks like ATLS is not using certificate validation context. How do we tell whether ATLS config is meant for mutual TLS?

Comment (lizan) [reply include/envoy/network/transport_socket.h]: did you mean ALTS? If you expose TransportSocket proto does that work?

**Pull request closed without merging**